api/resources_portal/management/commands/material_additional_metadata_key_change.py

```diff
@@ -0,0 +1,183 @@
+            current_slice = query_set[current_page : current_page + (size - current_page)]
+            current_page += size - current_page
+        else:
+            current_slice = query_set[current_page : current_page + pagination]
+            current_page += pagination
+
+        current_materials = []
+        for material in current_slice:
+            current_material = function(material)
+            current_material.updated_at = timezone.now()
+            current_materials.append(current_material)
+
+        Material.objects.bulk_update(current_materials, ["additional_metadata", "updated_at"])
+        materials_updated += len(current_materials)
+        logger.info(
+            f"{len(current_materials)} materials processed. {materials_updated} total processed so far."
+        )
+
+    return materials_updated
+
+
+def log_nonapplicable_materials(query_set, key):
+    if len(query_set) > 0:
+        logger.warning(
+            f"The key '{key}' was not present in {len(query_set)} materials. The IDs of these materials are: "
+        )
+
+        idString = ""
+        for material in query_set:
+            idString += str(material.id) + ", "
+
+        idString = idString.strip(", ")
+        logger.warning(idString)
+
+
+def add_key(category, new_key, default, pagination):
+    query_set = None
+    materials_changed = 0
+
+    if category:
+        query_set = Material.objects.filter(category=category)
+    else:
+        query_set = Material.objects.all()
+
+    def add_func(material):
+        material.additional_metadata[new_key] = default
+        return material
+
+    materials_changed = paginate_queryset(query_set, add_func, pagination)
+
+    logger.info(
+        f"The key '{new_key}' was added to {materials_changed} materials. The default value was '{default}'."
+    )
+
+
+def remove_key(category, old_key, pagination):
+    query_set = None
+    materials_changed = 0
+
+    if category:
+        query_set = Material.objects.filter(category=category)
+    else:
+        query_set = Material.objects.all()
+
+    filtered_query_set = query_set.filter(additional_metadata__has_key=old_key)
+    not_applicable_query_set = query_set.exclude(additional_metadata__has_key=old_key)
+
+    # Evaluates the queryset. Querysets are usually lazy, this forces evaluation
+    len(not_applicable_query_set)
+
+    def add_func(material):
+        material.additional_metadata.pop(old_key)
+        return material
+
+    materials_changed = paginate_queryset(filtered_query_set, add_func, pagination)
+
+    logger.info(f"The key '{old_key}' was removed from {materials_changed} materials.")
+
+    log_nonapplicable_materials(not_applicable_query_set, old_key)
+
+
+def material_additional_metadata_key_change(category, old_key, new_key, pagination):
+    query_set = None
+    materials_changed = 0
+
+    if category:
+        query_set = Material.objects.filter(category=category)
+    else:
+        query_set = Material.objects.all()
```

Comment:
This is repeated in each of these functions. Why not have this function just take the queryset as a parameter, so you can do this filtering once. If you wanted, you could even just have it take a pre-paginated queryset, although that'd be a bigger refactor.

Reply:
Good idea. I moved the category parsing into the main function body.